### PR TITLE
8256643: Terminally deprecate ThreadGroup stop, destroy, isDestroyed, setDaemon and isDaemon

### PR DESCRIPTION
**Pull request by AlanBateman**
This change terminally deprecates the following methods defined by java.lang.ThreadGroup 

- stop 
- destroy 
- isDestroyed 
- setDaemon 
- isDaemon 

The stop method has been deprecated since=1.2 because it is inherently unsafe. It is time to terminally deprecate this method so it can be removed in a future release. Thread.stop will be examined in a separate issue. 

The destroy, isDestroyed, setDaemon, isDaemon methods support the mechanism to explicitly or automatically destroy a thread group. As detailed in JDK-8252885, the mechanism to destroy thread groups is flawed and racy. Furthermore, this mechanism inhibits efforts to drop the reference from a thread group to its threads (so that thread creation, starting and termination do not need to coordinate with their thread group). These methods should be terminally deprecated so they can be degraded in a future release and eventually removed.

CSR with more information:  https://bugs.openjdk.java.net/browse/JDK-8256644

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (6/6 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8256643](https://bugs.openjdk.java.net/browse/JDK-8256643): Terminally deprecate ThreadGroup stop, destroy, isDestroyed, setDaemon and isDaemon

### Reviewers
 * [Sergey Bylokhov](https://openjdk.java.net/census#serb) (@mrserb - **Reviewer**) ⚠️ Review applies to ca656ba129153d409c0d8388aec2413f3b35e248
 * [Roger Riggs](https://openjdk.java.net/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to ca656ba129153d409c0d8388aec2413f3b35e248
 * [Iris Clark](https://openjdk.java.net/census#iris) (@irisclark - **Reviewer**) ⚠️ Review applies to ca656ba129153d409c0d8388aec2413f3b35e248
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to ca656ba129153d409c0d8388aec2413f3b35e248
 * [Stuart Marks](https://openjdk.java.net/census#smarks) (@stuart-marks - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1318/head:pull/1318`
`$ git checkout pull/1318`

### CODE REVIEW

**Review by mrserb — APPROVED** [review]

**Comment by mrserb** [comment]
The changes in the awt are trivial and look fine.

**Comment by seanjmullan** [comment]
I think it would be useful in the javadoc of the RuntimePermission targets (stopThread, etc) to add a note/link that the corresponding method that the permission applies to is terminally deprecated. Something as simple as "Note that the Thread.stop() method is terminally deprecated and will be deprecated in a future release."

File: src/java.base/share/classes/java/lang/ThreadGroup.java [hunk 1]
```diff
@@ -185,7 +185,14 @@ public final int getMaxPriority() {
      * @return  {@code true} if this thread group is a daemon thread group;
      *          {@code false} otherwise.
      * @since   1.0
+     *
+     * @deprecated Destroying a thread group and its subgroups is not an atomic
+     *             operation and is inherently racy with respect to threads that
+     *             have been created but not have started. The concept of daemon
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/lang/ThreadGroup.java:191]: "a daemon group" sounds better to me.

File: src/java.base/share/classes/java/lang/ThreadGroup.java [hunk 2]
```diff
@@ -195,7 +202,14 @@ public final boolean isDaemon() {
      *
      * @return  true if this object is destroyed
      * @since   1.1
+     *
+     * @deprecated Destroying a thread group and its subgroups is not an atomic
+     *             operation and is inherently racy with respect to threads that
+     *             have been created but not have started. The ability to explicitly
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/lang/ThreadGroup.java:208]: "have not started"

File: src/java.base/share/classes/java/lang/ThreadGroup.java [hunk 3]
```diff
@@ -217,7 +231,14 @@ public synchronized boolean isDestroyed() {
      * @see        java.lang.SecurityException
      * @see        java.lang.ThreadGroup#checkAccess()
      * @since      1.0
+     *
+     * @deprecated Destroying a thread group and its subgroups is not an atomic
+     *             operation and is inherently racy with respect to threads that
+     *             have been created but not have started. The concept of daemon
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/lang/ThreadGroup.java:237]: same comment as above - "a daemon group"

File: src/java.base/share/classes/java/lang/ThreadGroup.java [hunk 4]
```diff
@@ -763,7 +784,13 @@ public final void resume() {
      *               thread group.
      * @see        java.lang.ThreadGroup#checkAccess()
      * @since      1.0
+     *
+     * @deprecated Destroying a thread group and its subgroups is not an atomic
+     *             operation and is inherently racy with respect to threads that
+     *             have been created but not have started. The ability to explicitly
```
Comment (seanjmullan) [inline src/java.base/share/classes/java/lang/ThreadGroup.java:790]: "have not started"

File: src/java.base/share/classes/java/lang/ThreadGroup.java (hunk 1 above)
Comment (AlanBateman) [reply src/java.base/share/classes/java/lang/ThreadGroup.java:191]: The existing spec uses "daemon thread group" so I'd prefer to stick with that.

File: src/java.base/share/classes/java/lang/ThreadGroup.java (hunk 2 above)
Comment (AlanBateman) [reply src/java.base/share/classes/java/lang/ThreadGroup.java:208]: Thanks, there's a typo there.

**Comment by AlanBateman** [comment]
> I think it would be useful in the javadoc of the RuntimePermission targets (stopThread, etc) to add a note/link that the corresponding method that the permission applies to is terminally deprecated. Something as simple as "Note that the Thread.stop() method is terminally deprecated and will be removed in a future release."

We haven't change Thread.stop here, I would like to get that done too but as separate issue because it will require wider discussion. The main thing for now is the prep work before proposing to shove ThreadGroup to the side.

**Comment by seanjmullan** [comment]
> > I think it would be useful in the javadoc of the RuntimePermission targets (stopThread, etc) to add a note/link that the corresponding method that the permission applies to is terminally deprecated. Something as simple as "Note that the Thread.stop() method is terminally deprecated and will be removed in a future release."
> 
> We haven't changed Thread.stop here, I would like to get that done too but as separate issue because it will require wider discussion. The main thing for now is the prep work before proposing to shove ThreadGroup to the side.

Ok, but then how about putting a similar note in the javadoc for the RuntimePermission "modifyThreadGroup" target?

File: src/java.base/share/classes/java/lang/ThreadGroup.java [hunk 5]
```diff
@@ -185,7 +185,14 @@ public final int getMaxPriority() {
      * @return  {@code true} if this thread group is a daemon thread group;
      *          {@code false} otherwise.
      * @since   1.0
+     *
+     * @deprecated Destroying a thread group and its subgroups is not an atomic
+     *             operation and is inherently racy with respect to threads that
+     *             have been created but have not started. The concept of daemon
+     *             thread group that is automatically destroyed will be removed
+     *             in a future release.
      */
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/lang/ThreadGroup.java:194]: I would switch the order of the sentences.  Put the action first and the rationale second. $.02
(in all the @ deprecated javadoc).

**Review by RogerRiggs — APPROVED** [review]

**Comment by AlanBateman** [comment]
> Ok, but then how about putting a similar note in the javadoc for the RuntimePermission "modifyThreadGroup" target?

The "modifyThread" and "modifyThreadGroup" permission targets list methods that have been terminally deprecated for some time, are you looking for both permission targets to be updated? I could imagine doing an overall here, also re-visiting SecurityManager checkAccess(ThreadGroup) but it feels like it's beyond the scope of the deprecation proposed here and would be better off being left to when we eventually degrade and/or remove these methods.

**Comment by seanjmullan** [comment]
> > Ok, but then how about putting a similar note in the javadoc for the RuntimePermission "modifyThreadGroup" target?
> 
> The "modifyThread" and "modifyThreadGroup" permission targets list methods that have been terminally deprecated for some time, are you looking for both permission targets to be updated? I could imagine doing an overall here, also re-visiting SecurityManager checkAccess(ThreadGroup) but it feels like it's beyond the scope of the deprecation proposed here and would be better off being left to when we eventually degrade and/or remove these methods.

Ok, I see there is a broader context for my comment. I think then it makes sense to open a separate issue to specify these various RuntimePermission targets that are associated with deprecated APIs. In my mind, these permission targets are standard targets and should also be "deprecated" even if there is no annotation that can formally indicate that. For comparison, SecurityPermission puts these in a separate table in the class summary: https://download.java.net/java/early_access/jdk16/docs/api/java.base/java/security/SecurityPermission.html
If you agree, I can file an issue.

**Comment by AlanBateman** [comment]
> If you agree, I can file an issue.

Yes, make sense to separate this out, esp. permission targets such as "stopThread" where all usages are in deprecated methods. However, I don't expect "modifyThreadGroup" would be deprecated, at least not yet, because there are usages in several methods that are not deprecated.

**Review by mrserb — APPROVED** [review]

**Review by irisclark — APPROVED** [review]

**Review by mlchung — APPROVED** [review]

**Comment by stuart-marks** [comment]
I think the current deprecation wording is actually too specific regarding the raciness between TG destruction and created-but-not-started threads. That's just one of the flaws of thread groups. In fact, I think there are enough weirdnesses and race conditions around all destruction-related operations of thread groups that the whole concept is fundamentally flawed. We should just say that. How about this:

> ThreadGroup's destruction mechanisms are fundamentally flawed. Therefore, the ThreadGroup methods destroy(), isDestroyed(), setDaemon(), and isDaemon(), which relate to ThreadGroup destruction, have been deprecated and may be removed from a future version of the system.

I think there are too many subtle details to include a justification here about why TG destruction is fundamentally flawed, so we just have to assert that. Unfortunately the writeups in the JEP and CSR are in draft state so we can't link to them. Maybe when the JEP is published we can add a link to it from here later.

**Comment by AlanBateman** [comment]
Okay, I think I agree that the first sentence needs to be a bit more general so I've re-worded it. I used "inherently" rather than "fundamentally" to be consistent with the other deprecation text in Thread/ThreadGroup. If you are okay with the updated text then I'll transfer it to the CSR.

**Review by stuart-marks — APPROVED** [review]
I'm ok with the wording "inherently flawed."

**Pull request closed without merging**